lib/ts/recipe/totp/types.ts

```diff
@@ -0,0 +1,260 @@
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type RecipeInterface = {
+    createDevice: (input: {
+        userId: string;
+        userIdentifierInfo?: string;
+        deviceName?: string;
+        skew?: number;
+        period?: number;
+        userContext: UserContext;
+    }) => Promise<
+        | {
+              status: "OK";
+              deviceName: string;
+              secret: string;
+              qrCodeString: string;
+          }
+        | {
+              status: "DEVICE_ALREADY_EXISTS_ERROR";
+          }
+    >;
+    updateDevice: (input: {
+        userId: string;
+        existingDeviceName: string;
+        newDeviceName: string;
+        userContext: UserContext;
+    }) => Promise<{
+        status: "OK" | "UNKNOWN_DEVICE_ERROR" | "DEVICE_ALREADY_EXISTS_ERROR";
+    }>;
+    listDevices: (input: {
+        userId: string;
+        userContext: UserContext;
+    }) => Promise<{
+        status: "OK";
+        devices: {
+            name: string;
+            period: number;
+            skew: number;
+            verified: boolean;
+        }[];
+    }>;
+    removeDevice: (input: {
+        userId: string;
+        deviceName: string;
+        userContext: UserContext;
+    }) => Promise<{
+        status: "OK";
+        didDeviceExist: boolean;
+    }>;
+    verifyDevice: (input: {
+        tenantId: string;
+        userId: string;
+        deviceName: string;
+        totp: string;
+        userContext: UserContext;
+    }) => Promise<
+        | {
+              status: "OK";
+              wasAlreadyVerified: boolean;
+          }
+        | {
+              status: "UNKNOWN_DEVICE_ERROR";
+          }
+        | {
+              status: "INVALID_TOTP_ERROR";
+              currentNumberOfFailedAttempts: number;
+              maxNumberOfFailedAttempts: number;
+          }
+        | {
+              status: "LIMIT_REACHED_ERROR";
+              retryAfterMs: number;
+          }
+    >;
+    verifyTOTP: (input: {
+        tenantId: string;
+        userId: string;
+        totp: string;
+        userContext: UserContext;
+    }) => Promise<
+        | {
+              status: "OK" | "UNKNOWN_USER_ID_ERROR";
+          }
+        | {
+              status: "INVALID_TOTP_ERROR";
+              currentNumberOfFailedAttempts: number;
+              maxNumberOfFailedAttempts: number;
+          }
+        | {
+              status: "LIMIT_REACHED_ERROR";
+              retryAfterMs: number;
+          }
+    >;
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+export type APIInterface = {
+    createDevicePOST: (input: {
```

Comment:
All apis can also be undefined.. see APIInterface in other recipes!

Reply:
Done